## Segmenting and Clustering Neighborhoods in Toronto

In this assignment, we'll be clustering the neighborhoods in Toronto and will be doing the following:
1. Scraping the neighborhoods of Toronto from wikipedia
2. Retreive the latitude and longitude of each Postal Code
3. Use k-means to do a clustering on the neighborhoods

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

import folium
from geopy.geocoders import Nominatim 
import requests
import urllib.request
from bs4 import BeautifulSoup

print('Libraries imported.')

Libraries imported.


### Section 1: Scraping of Toronto neighborhoods

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
can_soup = BeautifulSoup(requests.get(url).text)

In [3]:
cells = []
for cell in can_soup.find('tbody').find_all("td"):
    cells.append(cell.text.strip())

In [4]:
cells[:5]

['M1ANot assigned',
 'M2ANot assigned',
 'M3ANorth York(Parkwoods)',
 'M4ANorth York(Victoria Village)',
 'M5ADowntown Toronto(Regent Park / Harbourfront)']

We observe that the first 3 letters contain the postal code, anything within the parentheses are neighborhoods and what remains between is the borough

In [5]:
can_soup.find('tbody').find_all("td")

[<td style="width:11%; vertical-align:top; color:#ccc;">
 <p><b>M1A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
 </p>
 </td>,
 <td style="width:11%; vertical-align:top; color:#ccc;">
 <p><b>M2A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
 </p>
 </td>,
 <td style="width:11%; vertical-align:top;">
 <p><b>M3A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>)</span>
 </p>
 </td>,
 <td style="width:11%; vertical-align:top;">
 <p><b>M4A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>)</span>
 </p>
 </td>,
 <td style="width:11%; vertical-align:top;">
 <p><b>M5A</b><br/><span style="font-size:85%;"><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a><br/>(<a href="/wiki/Regent_Park" titl

In [6]:
toronto_data = pd.DataFrame(columns=["PostalCode", "Borough", "Neighborhood"])
for cell in can_soup.find('tbody').find_all("td"):
    info = cell.text.strip()
    if "Not assigned" in info: 
        continue
    postal_code = info[:3]
    borough = (cell.span.text).split('(')[0]
    neighborhood = (((((cell.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
    toronto_data = toronto_data.append({'PostalCode':postal_code, 
                          'Borough':borough,
                          'Neighborhood':neighborhood}, ignore_index=True)

toronto_data['Borough']=toronto_data['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                                         'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                                         'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                                         'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})    
toronto_data.head()        

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

We see that the data has been beautifully scraped! However due to the way the table was created in wikipedia, rows are read first so the postalcode is going in the order of 'M3A' -> 'M4A' instead of 'M3A' -> 'M3B'. We will sort by postalcode and then we'll be done for section 1.

In [7]:
toronto_data.sort_values(by='PostalCode', axis=0, inplace=True)
toronto_data.head()

PostalCode      Borough                            Neighborhood
6         M1B  Scarborough                          Malvern, Rouge
12        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
18        M1E  Scarborough       Guildwood, Morningside, West Hill
22        M1G  Scarborough                                  Woburn
26        M1H  Scarborough                               Cedarbrae

In [8]:
toronto_data.shape

(103, 3)

### Section 2:  <strike>Using Geocoder to</strike> Retrieve the latitude and longitude of each Postal Code

In this section, we will be using the provided data instead. These coordinates will be utilized in section 3 to be fed into foursquare API to explore the neighborhood. The assignment mentioned to user Geocoder but unfortunately geocoder always returns none for me.

In [24]:
lat_lng_df = pd.read_csv("Geospatial_Coordinates.csv")
lat_lng_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [28]:
toronto_df = pd.merge(toronto_data,
                     lat_lng_df,
                     how='inner',
                     left_on='PostalCode',
                     right_on='Postal Code')
toronto_df.drop(columns='Postal Code', inplace=True)
toronto_df.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [29]:
toronto_df.shape

(103, 5)

### Section 3: Exploring and clustering neighborhoods in Toronto

We'll be selecting for the boroughs that contain the word 'Toronto'.

In [32]:
toronto_min_df = toronto_df[toronto_df['Borough'].str.contains("Toronto")]
toronto_min_df

PostalCode                 Borough  \
37        M4E            East Toronto   
40        M4J  East York/East Toronto   
41        M4K            East Toronto   
42        M4L            East Toronto   
43        M4M            East Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
46        M4R         Central Toronto   
47        M4S         Central Toronto   
48        M4T         Central Toronto   
49        M4V         Central Toronto   
50        M4W        Downtown Toronto   
51        M4X        Downtown Toronto   
52        M4Y        Downtown Toronto   
53        M5A        Downtown Toronto   
54        M5B        Downtown Toronto   
55        M5C        Downtown Toronto   
56        M5E        Downtown Toronto   
57        M5G        Downtown Toronto   
58        M5H        Downtown Toronto   
59        M5J        Downtown Toronto   
60        M5K        Downtown Toronto   
61        M5L        Downtown Toronto   
63        M5N         Central Toronto   
64        M5P         Central Toronto   
65        M5R         Central Toronto   
66        M5S        Downtown Toronto   
67        M5T        Downtown Toronto   
68        M5V        Downtown Toronto   
69        M5W  Downtown Toronto Stn A   
70        M5X        Downtown Toronto   
75        M6G        Downtown Toronto   
76        M6H            West Toronto   
77        M6J            West Toronto   
78        M6K            West Toronto   
82        M6P            West Toronto   
83        M6R            West Toronto   
84        M6S            West Toronto   
87        M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
40                                 The Danforth  East  43.685347 -79.338106  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64                           Forest Hill North & West  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
69                    

In [74]:
toronto_min_df.shape

(39, 5)

First we create a folium map of Toronto, Canada, this will act as the base for subsequent visualizations.

In [52]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="canada_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(f'The geograpical coordinate of {address} are {latitude}, {longitude}.')
map_canada = folium.Map(location=[latitude, longitude], zoom_start=11.5)
map_canada

The geograpical coordinate of Toronto, Canada are 43.6534817, -79.3839347.


We move on to add markers to map

In [53]:
for lat, lng, borough, neighborhood in zip(toronto_min_df['Latitude'], toronto_min_df['Longitude'], toronto_min_df['Borough'], toronto_min_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_canada)  
    
map_canada

Specifying Foursquare API Credentials, due to obvious reasons, the api ID, and secret are veiled in a separate file called config.

In [54]:
from config import credentials
CLIENT_ID = credentials['id']
CLIENT_SECRET = credentials['secret']
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value


In [55]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [56]:
toronto_venues = getNearbyVenues(names=toronto_min_df['Neighborhood'],
                                 latitudes=toronto_min_df['Latitude'],
                                 longitudes=toronto_min_df['Longitude']
                                )
toronto_venues.shape

(1483, 7)

In [57]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0  The Big Carrot Natural Food Market       43.678879       -79.297734   
1                   Glen Manor Ravine       43.676821       -79.293942   
2                   Glen Stewart Park       43.675278       -79.294647   
3                 Grover Pub and Grub       43.679181       -79.297215   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0  Health Food Store  
1              Trail  
2               Park  
3                Pub  
4       Neighborhood

In [58]:
toronto_venues[['Neighborhood','Venue']].groupby('Neighborhood').count()

Venue
Neighborhood                                             
Berczy Park                                            47
Brockton, Parkdale Village, Exhibition Place           24
CN Tower, King and Spadina, Railway Lands, Harb...     14
Central Bay Street                                     63
Christie                                               14
Church and Wellesley                                   68
Commerce Court, Victoria Hotel                        100
Davisville                                             26
Davisville North                                        8
Dufferin, Dovercourt Village                           14
Enclave of M4L                                         16
Enclave of M5E                                         85
First Canadian Place, Underground city                100
Forest Hill North & West                                4
Garden District, Ryerson                              100
Harbourfront East, Union Station, Toronto Islands     100
High Park, The Junction South                          17
India Bazaar, The Beaches West                         17
Kensington Market, Chinatown, Grange Park              57
Lawrence Park                                           3
Little Portugal, Trinity                               42
North Toronto West                                     15
Parkdale, Roncesvalles                                 12
Regent Park, Harbourfront                              45
Richmond, Adelaide, King                              100
Rosedale                                                4
Roselawn                                                3
Runnymede, Swansea                                     31
St. James Town                                         81
St. James Town, Cabbagetown                            35
Studio District                                        26
Summerhill West, Rathnelly, South Hill, Forest ...      9
The Annex, North Midtown, Yorkville                    16
The Beaches                                             6
The Danforth  East                                      2
The Danforth West, Riverdale                           43
Toronto Dominion Centre, Design Exchange              100
University of Toronto, Harbord                         36

In [60]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 216 uniques categories.


From above, we see that our search via Foursquare has returned us 1483 venues across the 39 neighborhoods. Among the venues, there are 216 unique categories. Next we move on to implementing K-means clustering. K-means clustering does not accept categorical values. As such, we'll use one-hot encoding on categories, turning it into 216 features corresponding to each unique category.

In [69]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Adult Boutique  Afghan Restaurant  Airport  Airport Food Court  \
0               0                  0        0                   0   
1               0                  0        0                   0   
2               0                  0        0                   0   
3               0                  0        0                   0   
4               0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                    0             0         0            0   
1                    0             0         0            0   
2                    0             0         0            0   
3                    0             0         0            0   
4                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  \
0                    0                 0              0          0   
1                    0                 0              0          0   
2                    0                 0              0          0   
3                    0                 0              0          0   
4                    0                 0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0                     0               0        0                0   
1                     0               0        0                0   
2                     0               0        0                0   
3                     0               0        0                0   
4                     0               0        0                0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  Cheese Shop  \
0                          0            0                     0            0   
1                          0            0                     0            0   
2                          0            0     

In [70]:

# add neighborhood column back to dataframe
# there is already one column called neighborhood in the dataframe, hence we use another name
toronto_onehot.insert(0, column='Official Neighborhood', value = toronto_venues['Neighborhood']) 

toronto_onehot.head()

Official Neighborhood  Adult Boutique  Afghan Restaurant  Airport  \
0           The Beaches               0                  0        0   
1           The Beaches               0                  0        0   
2           The Beaches               0                  0        0   
3           The Beaches               0                  0        0   
4           The Beaches               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  \
0                    0                 0              0          0   
1                    0                 0              0          0   
2                    0                 0              0          0   
3                    0                 0              0          0   
4                    0                 0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0                     0               0        0                0   
1                     0               0        0                0   
2                     0               0        0                0   
3                     0               0        0                0   
4                     0               0        0                0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  Cheese Shop  \
0                          0            0                     0          

In [71]:
toronto_onehot.shape

(1483, 217)

Next, we collapse the categories by neighborhood since our clustering is meant to be done on each neighborhood, reducing the values to be the frequency of occurence of each category. A value of 0.2 for a category would mean that of the venes within the neighborhood, 20% are of said category.

In [72]:
toronto_grouped = toronto_onehot.groupby('Official Neighborhood').mean().reset_index()
toronto_grouped.head()

Official Neighborhood  Adult Boutique  \
0                                        Berczy Park             0.0   
1       Brockton, Parkdale Village, Exhibition Place             0.0   
2  CN Tower, King and Spadina, Railway Lands, Har...             0.0   
3                                 Central Bay Street             0.0   
4                                           Christie             0.0   

   Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0                0.0  0.000000            0.000000      0.000000   
1                0.0  0.000000            0.000000      0.000000   
2                0.0  0.071429            0.071429      0.071429   
3                0.0  0.000000            0.000000      0.000000   
4                0.0  0.000000            0.000000      0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                  0.0   
1        0.000000         0.000000          0.000000                  0.0   
2        0.071429         0.214286          0.142857                  0.0   
3        0.000000         0.000000          0.000000                  0.0   
4        0.000000         0.000000          0.000000                  0.0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0     0.021277             0.000000               0.0   
1           0.0       0.0     0.000000             0.000000               0.0   
2           0.0       0.0     0.000000             0.000000               0.0   
3           0.0       0.0     0.000000             0.015873               0.0   
4           0.0       0.0     0.000000             0.000000               0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery      Bank  \
0            0.0        0.0    0.000000    0.021277  0.063830  0.000000   
1            0.0        0.0    0.000000    0.000000  0.083333  0.000000   
2            0.0        0.0    0.000000    0.000000  0.000000  0.000000   
3            0.0        0.0    0.000000    0.000000  0.000000  0.031746   
4            0.0        0.0    0.071429    0.000000  0.000000  0.000000   

        Bar  Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  \
0  0.000000               0.0            0.021277  0.021277              0.0   
1  0.041667               0.0            0.000000  0.000000              0.0   
2  0.000000               0.0            0.000000  0.000000              0.0   
3  0.015873               0.0            0.000000  0.000000              0.0   
4  0.000000               0.0            0.000000  0.000000              0.0   

   Beer Bar  Beer Store  Belgian Restaurant    Bistro  Board Shop  \
0  0.042553         0.0                 0.0  0.021277         0.0   
1  0.000000         0.0                 0.0  0.000000         0.0   
2  0.000000         0.0                 0.0  0.000000         0.0   
3  0.000000         0.0                 0.0  0.000000         0.0   
4  0.000000         0.0                 0.0  0.000000         0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0       0.000000        0.0       0.0                   0.0        0.000000   
1       0.000000        0.0       0.0                   0.0        0.083333   
2       0.071429        0.0       0.0                   0.0        0.000000   
3       0.000000        0.0       0.0                   0.0        0.000000   
4       0.000000        0.0       0.0                   0.0        0.000000   

   Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line   Butcher  \
0      0.0         0.000000      0.000000            0.0       0.0  0.021277   
1      0.0         0.000000      0.000000            0.0       0.0  0.000000   
2      0.0         0.000000      0.000000            0.0       0.0  0.000000   
3      0.0         0.015873      0.031746            0.0       0.0  0.000000   
4      0.0       

In [73]:
toronto_grouped.shape

(38, 217)

There are too many features to take note of, hence we attempt to condense them into just 10 of the most common venues in each neighbor hood

In [79]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [80]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Official Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Official Neighborhood'] = toronto_grouped['Official Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Official Neighborhood 1st Most Common Venue  \
0                                        Berczy Park          Cocktail Bar   
1       Brockton, Parkdale Village, Exhibition Place                Bakery   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery           Coffee Shop        Sandwich Place   
1        Breakfast Spot        Sandwich Place                  Café   
2      Airport Terminal       Harbor / Marina   Rental Car Location   
3        Sandwich Place      Sushi Restaurant                  Café   
4                  Café                  Park            Restaurant   

  5th Most Common Venue 6th Most Common Venue          7th Most Common Venue  \
0              Beer Bar    Seafood Restaurant  Vegetarian / Vegan Restaurant   
1           Coffee Shop    Italian Restaurant         Furniture / Home Store   
2      Sculpture Garden         Boat or Ferry                 Airport Lounge   
3    Italian Restaurant   Japanese Restaurant                   Burger Joint   
4            Baby Store           Coffee Shop             Italian Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market   Sporting Goods Shop             Steakhouse  
1          Climbing Gym                   Bar             Restaurant  
2          Airport Gate               Airport     Airport Food Court  
3           Salad Place            Restaurant            Pizza Place  
4             Nightclub   Moroccan Restaurant                 Museum

Now, we are ready to run k-means clustering to sort the neighborhood into 3 clusters

In [98]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Official Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=15, n_init=30).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [102]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

neighborhoods_venues_sorted.rename(columns = {'Official Neighborhood': 'Neighborhood'}, inplace=True)
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                                        Berczy Park   
1               0       Brockton, Parkdale Village, Exhibition Place   
2               0  CN Tower, King and Spadina, Railway Lands, Har...   
3               0                                 Central Bay Street   
4               0                                           Christie   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0          Cocktail Bar                Bakery           Coffee Shop   
1                Bakery        Breakfast Spot        Sandwich Place   
2       Airport Service      Airport Terminal       Harbor / Marina   
3           Coffee Shop        Sandwich Place      Sushi Restaurant   
4         Grocery Store                  Café                  Park   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Sandwich Place              Beer Bar    Seafood Restaurant   
1                  Café           Coffee Shop    Italian Restaurant   
2   Rental Car Location      Sculpture Garden         Boat or Ferry   
3                  Café    Italian Restaurant   Japanese Restaurant   
4            Restaurant            Baby Store           Coffee Shop   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Vegetarian / Vegan Restaurant        Farmers Market   Sporting Goods Shop   
1         Furniture / Home Store          Climbing Gym                   Bar   
2                 Airport Lounge          Airport Gate               Airport   
3                   Burger Joint           Salad Place            Restaurant   
4             Italian Restaurant             Nightclub   Moroccan Restaurant   

  10th Most Common Venue  
0             Steakhouse  
1             Restaurant  
2     Airport Food Court  
3            Pizza Place  
4                 Museum

In [104]:
toronto_merged = toronto_min_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode                 Borough                    Neighborhood  \
37        M4E            East Toronto                     The Beaches   
40        M4J  East York/East Toronto              The Danforth  East   
41        M4K            East Toronto    The Danforth West, Riverdale   
42        M4L            East Toronto  India Bazaar, The Beaches West   
43        M4M            East Toronto                 Studio District   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
37  43.676357 -79.293031             0.0     Health Food Store   
40  43.685347 -79.338106             1.0                  Park   
41  43.679557 -79.352188             0.0      Greek Restaurant   
42  43.668999 -79.315572             0.0        Sandwich Place   
43  43.659526 -79.340923             0.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
37                 Trail      Asian Restaurant                   Pub   
40     Convenience Store        Adult Boutique                Museum   
41           Coffee Shop    Italian Restaurant      Sushi Restaurant   
42  Fast Food Restaurant            Board Shop            Steakhouse   
43                Bakery                  Café             Gastropub   

        5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
37               Neighborhood                  Park            Optical Shop   
40   Mediterranean Restaurant           Men's Store      Mexican Restaurant   
41             Ice Cream Shop           Yoga Studio  Furniture / Home Store   
42                       Park           Pizza Place                     Pub   
43  Middle Eastern Restaurant                  Food             Cheese Shop   

        8th Most Common Venue      9th Most Common Venue  \
37              Movie Theater   Mediterranean Restaurant   
40  Middle Eastern Restaurant         Miscellaneous Shop   
41             Sandwich Place                        Pub   
42                 Restaurant              Movie Theater   
43           Stationery Store  Latin American Restaurant   

        10th Most Common Venue  
37                 Men's Store  
40  Modern European Restaurant  
41          Frozen Yogurt Shop  
42               Burrito Place  
43                Neighborhood

In [114]:
toronto_merged = toronto_merged.dropna(axis=0)
toronto_merged.shape

(38, 16)

Finally, let's visualize the resulting clusters

In [115]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11.5)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Cluster 0

In [116]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
37            East Toronto             0.0     Health Food Store   
41            East Toronto             0.0      Greek Restaurant   
42            East Toronto             0.0        Sandwich Place   
43            East Toronto             0.0           Coffee Shop   
45         Central Toronto             0.0        Breakfast Spot   
46         Central Toronto             0.0        Clothing Store   
47         Central Toronto             0.0           Pizza Place   
49         Central Toronto             0.0           Coffee Shop   
51        Downtown Toronto             0.0           Coffee Shop   
52        Downtown Toronto             0.0   Japanese Restaurant   
53        Downtown Toronto             0.0           Coffee Shop   
54        Downtown Toronto             0.0           Coffee Shop   
55        Downtown Toronto             0.0           Coffee Shop   
56        Downtown Toronto             0.0          Cocktail Bar   
57        Downtown Toronto             0.0           Coffee Shop   
58        Downtown Toronto             0.0           Coffee Shop   
59        Downtown Toronto             0.0           Coffee Shop   
60        Downtown Toronto             0.0           Coffee Shop   
61        Downtown Toronto             0.0           Coffee Shop   
65         Central Toronto             0.0        Sandwich Place   
66        Downtown Toronto             0.0                  Café   
67        Downtown Toronto             0.0                  Café   
68        Downtown Toronto             0.0       Airport Service   
69  Downtown Toronto Stn A             0.0        Sandwich Place   
70        Downtown Toronto             0.0           Coffee Shop   
75        Downtown Toronto             0.0         Grocery Store   
76            West Toronto             0.0           Music Venue   
77            West Toronto             0.0                   Bar   
78            West Toronto             0.0                Bakery   
82            West Toronto             0.0                   Bar   
83            West Toronto             0.0           Coffee Shop   
84            West Toronto             0.0           Coffee Shop   
87   East Toronto Business             0.0    Light Rail Station   

            2nd Most Common Venue      3rd Most Common Venue  \
37                          Trail           Asian Restaurant   
41                    Coffee Shop         Italian Restaurant   
42           Fast Food Restaurant                 Board Shop   
43                         Bakery                       Café   
45                           Park       Gym / Fitness Center   
46                    Coffee Shop                 Restaurant   
47                 Sandwich Place           Sushi Restaurant   
49                    Pizza Place         Light Rail Station   
51                     Restaurant                       Café   
52               Sushi Restaurant                Coffee Shop   
53                           Park                     Bakery   
54                 Clothing Store             Sandwich Place   
55                           Café         Italian Restaurant   
56                         Bakery                Coffee Shop   
57                 Sandwich Place           Sushi Restaurant   
58                           Café             Sandwich Place   
59                           Café                      Hotel   
60                           Café                      Hotel   
61                           Café             Sandwich Place   
65                    Coffee Shop                       Café   
66                    Coffee Shop             Sandwich Place   
67  Vegetarian / Vegan Restaurant                Coffee Shop   
68               Airport Terminal            Harbor / Marina   
69                    Coffee Shop               Cocktail Bar   
70                 Sandwich Place                       Café   
75                           Café         

Cluster 1

In [117]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
40  East York/East Toronto             1.0                  Park   
44         Central Toronto             1.0                  Park   
50        Downtown Toronto             1.0                  Park   
64         Central Toronto             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
40     Convenience Store        Adult Boutique                Museum   
44              Bus Line           Swim School        Adult Boutique   
50            Playground                 Trail        Adult Boutique   
64                 Trail         Jewelry Store      Sushi Restaurant   

       5th Most Common Venue     6th Most Common Venue  \
40  Mediterranean Restaurant               Men's Store   
44             Movie Theater  Mediterranean Restaurant   
50             Movie Theater       Martial Arts School   
64            Adult Boutique       Moroccan Restaurant   

       7th Most Common Venue      8th Most Common Venue  \
40        Mexican Restaurant  Middle Eastern Restaurant   
44               Men's Store         Mexican Restaurant   
50  Mediterranean Restaurant                Men's Store   
64       Martial Arts School   Mediterranean Restaurant   

        9th Most Common Venue      10th Most Common Venue  
40         Miscellaneous Shop  Modern European Restaurant  
44  Middle Eastern Restaurant          Miscellaneous Shop  
50         Mexican Restaurant   Middle Eastern Restaurant  
64                Men's Store          Mexican Restaurant

Cluster 2

In [118]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels    1st Most Common Venue  \
63  Central Toronto             2.0  Health & Beauty Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
63          Home Service                Garden           Music Venue   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63  Mediterranean Restaurant           Men's Store    Mexican Restaurant   

        8th Most Common Venue 9th Most Common Venue  \
63  Middle Eastern Restaurant    Miscellaneous Shop   

        10th Most Common Venue  
63  Modern European Restaurant

### Conclusion

From above, it seems like we have 3 main clusters. Through the second one is so unique, it is in a cluster by itself and might seem like an outlier. The first cluster is primarily popular areas for coffee shop, sandwich, pizza places. Second cluster is a popular park area.